### DISCUSSION TITLE
[tinytex] TinyTeX: A lightweight, cross-platform, portable, and easy-to-maintain LaTeX distribution based on TeX Live | /tinytex/

### DISCUSSION THREAD

**Discussion by yihui** [rbind/yihui#278] [Comments]
https://yihui.org/tinytex/

**Comment by shrektan** [comment]
公司服务器没有互联网链接，最近两天鼓捣了会TexLive，想看看怎么才能弄一个离线的Repo。然后，我发现真是太乱了…真看不懂…

**Reply by yihui** [reply]
虽说离线 repo 已无必要（装我编好的 TinyTeX-2 就对了），但正好前段时间研究了一下这个问题，就来分享个答案。其实也并不难，核心就是用 `rsync` 去把这个目录 `rsync://rsync.dante.ctan.org/CTAN/systems/texlive/tlnet/` 同步到本地，然后本地这个文件夹就可以当 TeX Live 的仓库用了，所有需要的东西全在这一个文件夹中，例如 `tlmgr install` 安装包就可以从这个文件夹的 archive/ 子文件夹中获得包。

我研究这个是为了解决一个多年来的痛点，就是 CTAN 没有一个基于 CDN 云端分发的镜像，而各个 CTAN 镜像同步又不一致，有的早有的晚，这就导致所有用户跑起 `tlmgr` 命令来不能保证大家在同样的时间都能得到同样的结果（取决于设置的镜像）。这问题在多数时候也不是个大事，但 TeX Live 有个很要命的年度痛点，就是每年它放出新的年度版本之后，前一年的 `tlmgr` 基本就废了，用户无法用旧的 `tlmgr` 从新的 CTAN 镜像中装包，必须要重装新版本的 TeX Live。而这个破事的麻烦之处在于，新版本的 TeX Live 放出来之后各镜像同步的时间和进度不一致；如果用户说好吧一咬牙装了新的 TeX Live，而他用的镜像还是个旧的，那他又将遇到反方向的问题，即新版本的 tlmgr 无法从旧仓库中装包。

我在研究了一番之后发现（1）CTAN 上大部分文件都是垃圾，整整几十 G 的垃圾，说它们是垃圾不是说质量（里面还有高德纳大人的一些历史旧作），而是那些文件真的是特么八辈子都不会有一个人去下载，而唯一有用的只有这一个 tlnet 文件夹；（2）这一个 tlnet 下也有大量垃圾，像什么他们辛辛苦苦为 `i386-netbsd` `x86_64-solaris` 这些根本没鸟人用的操作系统特意编译的二进制包（所有需要编译为二进制包的 LaTeX 包基本上都会为 15 种操作系统编 15 份），还有源代码、文档包等等，经过一清理，我发现所有文件加起来才八千个、2.3 G。

于是乎，本嬴政就果断出手了，用 Cloudflare R2 做了个文件仓库用来同步这个 tlnet 文件夹：https://tlnet.yihui.org/index.html 它是用 Cloudflare 的 CDN 分发的，所有可以全球保障速度；然后用 Github Action 实现日更：https://github.com/yihui/ctan-tlnet 这下终于解决了各大 CTAN 散户镜像的不同步问题了。现在 TinyTeX 就是从这个仓库构建，对用户默认也是从这里装包。

这一切基本上都是从 Claude 的网页对话框里聊天聊出来的，哈哈。基本都是它在向我输入，让我知道了好多我以前不懂的东西，而我给它最大的人工输入大概就是我知道哪些是垃圾。所有用到的这些服务全部都免费（R2 有操作限额，但八千个文件远远达不到免费的限额）。

**Reply by shrektan** [reply]
厉害，佩服佩服。看着就厉害，你还自己去研究了CTAN里面的东东，垃圾山里面去淘宝，哈哈。

不过我已经提前从源头解决了问题（把这些自动化报告的活甩出去了，哈哈，最好的工作方法是不做工作）。现在就还剩下些存量的报告了，然后我的方法是，构建一个docker image，然后我人肉一个个把我需要的tex包识别出来装进去，然后再也不改了。既然你这个只有2.3G,我下次切换到你这个源来试试，哈哈。

去年试了下 typst 感觉比latex斜杠看起来舒服点，不过现在没有这些自动化报告工作了，对于临时性的工作相关的报告，我基本就是生成markdown，然后用我弄的个网页app https://formaldoc.app 一秒转成公文格式的word文档。

我现在只能习惯于在CLI里面直接Claude Code或者用Codex.app进行聊天，聊好了直接让它干活。自己再手写一堆代码是绝无可能了…哪怕是AI解决不了的，我也用给它清晰地描述出来应该怎么写，然后它就噼里啪啦弄完了。你看看我最近的commits，差不多5000个…… 我已经上线了一个ios APP，明后天应该还能上线一个，哈哈。我现在是买了个Mac Mini放家里，然后手机上用termius，想起来啥了就用typeless语音转文字让AI干一堆活，再自动推到我手机来，我有空了就测试下，再反馈，哈哈。

PS. IOS APP开发也是一堆坑，如果没有AI，这个工作量是绝无可能完成…… MVP基本几个小时弄完，弄成真正的产品，打磨功能还有解决恶心的bug，一言难尽。

**Reply by yihui** [reply]
这御风飞行的速度，怕是过两天都要记不住几天前干了一票什么大事了，哈哈。感觉现在就像寒武纪生命大爆发一样，搓软件应用的速度就像搓丸子。

**Reply by shrektan** [reply]
确实，真正能记住的并坚持下来迭代推进的，才是“真”需求。

我的基本哲学是，起码得要有我自己一个用户，顺带弄成产品看看其他人要不要。现在软件开发成本极速降低了，终于可以好好为自己服务了。待看看寒武纪的软件应用多少能经历时间的蹉跎吧，哈哈。

**Comment by yihui** [comment]
这个你可以早一点问我，就不必浪费这个时间了。其实我已经把 TeX Live 全集都打包好了，只不过我还没昭告天下而已（因为我还没确定最终把这些超级大包存在哪里，是放我个人 Github 账号里还是转移到公司账号下）。全集是 TinyTeX-2：https://github.com/yihui/tinytex-releases/releases/tag/daily

通过 R 安装：

```r
tinytex:::install_prebuilt('TinyTeX-2')
```

或者通过脚本安装皆可：https://yihui.org/tinytex/#installation 设置环境变量 `TINYTEX_INSTALLER=TinyTeX-2`。

当然，你也可以直接下载文件，解压缩到合适的位置。

这个精简版的全集体积比官方应该是能小一半的。装它之后就不必再担心缺包了。

**Comment by shrektan** [comment]
本以为这个需求能很快找到答案，我还是低估了LaTeX的复杂程度！太赞了，我马上试试，谢谢！

**Comment by yihui** [comment]
嗯，架设一个可以离线使用的仓库估计是会有点复杂，但打包所有的 LaTeX 包就简单多了。既然架设库的主要目的是为了装缺失的包，那还不如就整锅端下来后一劳永逸了。

**Comment by shrektan** [comment]
是的。其实我本来就是想全装，但tlmgr install scheme-full 在最后一步老是报错，原因不详，已放弃。现在直接一步到位，非常开心，正合我意。

**Comment by yihui** [comment]
嗯，是的，这个命令确实是有出错的可能，我也一样读不懂错误消息、不明白原因。比如在 AppVeyor 平台上的 macOS 系统就会出错，我直接加个 `|| true` 给暴力忽略了错误：https://github.com/yihui/tinytex-releases/blob/4149573c79eae0d0d722218e0420d799f2c76cc1/appveyor.yml#L96 我试过在 Github Actions 的三大操作系统上运行也会出错，同样原因不详，所以现在我依然在用 AppVeyor 打包。

**Comment by Teddy-van-Jerry** [comment]
您好，我现在正在做一个仿真软件（[mmCEsim](https://github.com/mmcesim)）可以自动导出并编译 LaTeX 报告，仅在软件安装包里给用户提供 minimal 的 packages。我想是不是可以直接利用您在 `tools` 里的代码配合上 GitHub Actions 直接进行 build 呢？不确定是不是修改 `tools/pkgs-custom.txt` 还是 `tools/pkgs-yihui.txt`，再调整对应的 `build.yml` 内容即可实现 :-)

**Comment by yihui** [comment]
@Teddy-van-Jerry 是的，不过更简单的办法是你先安装 [`TinyTeX-0`](https://github.com/rstudio/tinytex-releases#releases)，在此基础上再用 `tlmgr install` 添加你需要的其它包，最后将 TinyTeX 文件夹打包发行。当然，用你说的办法从头编译也是可以的，只是稍微啰嗦一点。

**Comment by Teddy-van-Jerry** [comment]
哦哦明白啦，非常感谢！

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
See footnote 2 above. It seems ~/bin is not on your PATH variable.

You should not use sudo unless you are a system admin and install TinyTeX for other users in the system.

> Originally posted on 2017-12-16 15:13:09

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
How did you use TinyTeX? Are you compiling a LaTeX document or an R Markdown document? If the latter, all you need is

devtools::install_github(c('yihui/tinytex', 'rstudio/rmarkdown'))

If the former, you need to figure out the actual LaTeX package name:

$ tlmgr search --global --file pifont  
tlmgr: package repository <http://mirror.utexas.edu/ctan/systems/texlive/tlnet> (verified)  
psnfss:  
texmf-dist/tex/latex/psnfss/pifont.sty  
tex4ht:  
texmf-dist/tex/generic/tex4ht/pifont.4ht  
ucs:  
texmf-dist/doc/latex/ucs/config/pifont.ucf

And psnfss seems to be what you actually need.

> Originally posted on 2018-01-07 02:49:39

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Thanks for the feedback! I can surely fix this issue. May I ask why you wanted to install it to /usr/local/? Unless you were trying to install it for multiple users on the same Linux server, there is no benefit of installing TinyTeX to /usr/local/ as far as I can see.

> Originally posted on 2018-03-02 14:58:55

**Reply by yihui** [reply]
I see. That makes perfect sense! I'll fix it in the R package tinytex. In the meantime, I think you can use the mv command to move the directory, e.g.,

mv ～/.TinyTeX /usr/local/tinytex

Or wrap it in R: system('mv ～/.TinyTeX /usr/local/tinytex')

Then you need to run

/usr/local/tinytex/bin/*/tlmgr path add

> Originally posted on 2018-03-02 16:14:02

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Did you restart RStudio after installing tinytex? If you did, could you follow the instructions at <https://github.com/rstudio/bookdown/issues/536#issuecomment-367201781> and report the detailed error message? Thanks!

> Originally posted on 2018-03-05 19:16:16

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
The automatic installation of missing LaTeX packages currently only works for R Markdown documents. Rnw is not R Markdown. If you need that feature to work for Rnw documents, you have to call knitr::knit2pdf() manually to compile the Rnw document. Or wait for RStudio to support it: <https://github.com/yihui/tinytex/issues/10>

> Originally posted on 2018-03-14 14:27:36

**Reply by yihui** [reply]
Thanks! Your feedback was definitely very helpful!

> Originally posted on 2018-03-14 14:56:42

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
No, the batch file should be fully automated and does not require any edits. It should not pull the whole TeX Live, though. It should only install less than 70 LaTeX packages by default. If you are an R user, I recommend you to use the R function to install TinyTeX.

> Originally posted on 2018-03-19 19:35:33

**Reply by yihui** [reply]
You mean the R function tinytex::install_tinytex() didn't work?

There might be something wrong with your internet connection. I'm not sure. You can read the source code and go through these steps manually if the automated script doesn't work for you: <https://github.com/yihui/tinytex/blob/master/tools/install-windows.bat>

> Originally posted on 2018-03-22 18:43:37

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Did you restart RStudio? What does Sys.which('pdflatex') return in R? You can ignore the message "No TeX installation detected" if it does not return an empty string.

> Originally posted on 2018-03-26 20:40:07

**Reply by yihui** [reply]
Right. pdflatex is not on PATH. What does Sys.getenv("PATH") return? And tinytex:::texlive_root()? Does the directory tinytex:::win_app_dir('TinyTeX') exist?

> Originally posted on 2018-03-30 18:36:48

**Reply by yihui** [reply]
Can you check if the directory exists?

dir.exists("C:UsersvinceAppDataRoamingTinyTeX")

And can you run

system2("C:UsersvinceAppDataRoamingTinyTeXbinwin32tlmgr", c("path", "add"))

?

> Originally posted on 2018-04-05 22:15:33

**Reply by yihui** [reply]
That means installing TinyTeX failed. Could you retry tinytex::install_tinytex() and let me know the full output in the console?

> Originally posted on 2018-04-06 13:57:44

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Correct me if I'm wrong, but I guess you don't have to wait for the Debain package r-cran-tinytex, since tinytex is a pure R package, and easy to install with install.packages('tinytex') in R.

> Originally posted on 2018-03-28 22:09:38

**Reply by yihui** [reply]
The R package tinytex does not require TinyTeX (the TeX Live distribution I made): <https://yihui.name/tinytex/r/> knitr::knit2pdf() works with any LaTeX distributions.

I imagine you would have to do `apt-get install r-cran-knitr` somewhere. If you do that, I don't see a reason why you cannot do `Rscript -e "install.packages('tinytex')"`.

~~I don't know who builds the r-cran-knitr package on Debian, but you could request them to build r-cran-tinytex too. Actually the next version of rmarkdown will depend on tinytex, so hopefully you will be able to install r-cran-tinytex in a couple of months (at the moment, I don't have a plan to release the next version of rmarkdown to CRAN yet).~~ Actually r-cran-tinytex seems to be available on Debian: <https://packages.debian.org/source/sid/r-cran-tinytex>

> Originally posted on 2018-03-29 14:55:50

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
<https://github.com/yihui/tinytex/issues/24>

> Originally posted on 2018-05-10 13:44:04

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Without a minimal reproducible example, a blind guess is you need to use xelatex instead of pdflatex (perhaps with proper font settings). This is not an issue that TinyTeX can automatically solve.

> Originally posted on 2018-05-29 23:00:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I don't think this problem is specific to any particular LaTeX distribution. It is a general LaTeX problem. I don't know anything about Korean in LaTeX, but I guess you may have to use xelatex as the engine.

> Originally posted on 2018-07-01 20:50:54

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
The package name is psnfss.

> Originally posted on 2018-08-09 18:01:24

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
There seems to be something wrong with your internet connection. You may retry for a couple of times. If it still fails, please restart your R session and try this:

    devtools::install_github('yihui/tinytex')
    tinytex:::install_prebuilt()

> Originally posted on 2018-08-14 16:46:39

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
It is hard to tell what's going on without the full log (which you can obtain on [shinyapps.io](http://shinyapps.io)) or the source of your app.

> Originally posted on 2018-10-26 01:22:03

**Reply by yihui** [reply]
Okay. I replied at <https://github.com/yihui/knitr/issues/1521#issuecomment-433267757>

> Originally posted on 2018-10-26 02:42:08

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
How do you set the proxy for R?

For TeX Live, you can set the proxy info in ~/.wgetrc: <https://www.gnu.org/software/wget/manual/html_node/Proxies.html>

> Originally posted on 2018-11-05 15:36:45

**Reply by yihui** [reply]
Very glad to know that, and also congrats on your promotion!

> Originally posted on 2019-03-15 19:56:28

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
The warning message has been changed in the newer version of RStudio: <https://github.com/rstudio/rstudio/pull/3979> You may try the RStudio Preview version at the moment: <https://www.rstudio.com/products/rstudio/download/preview/>

> Originally posted on 2019-01-19 16:34:59

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Did you restart RStudio?

> Originally posted on 2019-01-22 00:02:48

**Reply by yihui** [reply]
That is not enough. I mean you should close RStudio and open it again.

> Originally posted on 2019-01-22 03:55:02

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Search for "Homebrew" in the FAQ: <https://yihui.name/tinytex/faq/>

> Originally posted on 2019-02-06 23:11:29

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Don't use the default pdflatex engine. Try xelatex instead: <https://bookdown.org/yihui/rmarkdown/pdf-document.html#advanced-customization-1>

> Originally posted on 2019-02-14 14:39:31

**Reply by yihui** [reply]
Please consider asking such questions on Stack Overflow with a reproducible example.

> Originally posted on 2019-02-18 14:51:45

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
If your R version is not the latest, please upgrade to the latest.

> Originally posted on 2019-02-25 04:50:42

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
At the moment, you can just ignore this message. When TeX Live 2019 becomes ready, run tinytex::reinstall_tinytex() to install it.

> Originally posted on 2019-03-04 04:31:17

**Reply by yihui** [reply]
My pleasure! Thanks for the kind words! :)

> Originally posted on 2019-03-04 19:11:57

**Reply by yihui** [reply]
TeX Live 2019 was *just* released about 12 hours ago. I guess you have to wait for a couple of days before it is fully propagated to most CTAN mirrors: <https://tug.org/pipermail/tex-live/2019-April/043607.html>

> Originally posted on 2019-04-30 17:28:21

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Hi Satyen,

1. I believe it is a Perl issue. I don't know anything about Perl or Perl modules, so I don't know how to install the Perl/Tk module.

2. I don't use Texmaker, and I don't know why it cannot find the "times" package. This package is indeed included in the default installation of TinyTeX, so I doubt if Texmaker has correctly found and used TinyTeX. Could it be using a certain leftover TeX Live installation?

3. Please read this web page again. In particular, the section "Maintenance": <https://yihui.name/tinytex/#maintenance> It seems you didn't understand the relationship between package file names and package names. For example, ocgbase is not a package name, but a filename in the ocgx2 package.

> Originally posted on 2019-06-21 03:55:00

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry, I cannot tell anything without a minimal reproducible example...

> Originally posted on 2019-08-13 22:48:29

**Reply by yihui** [reply]
Okay. Could you do me a favor to ask the question on Stack Overflow? I guess I know the answer but I don't have time to verify it, and I hope someone else could help. This issue has nothing to do with tinytex but Pandoc instead.

> Originally posted on 2019-08-14 02:59:19

**Reply by yihui** [reply]
Pandoc generated an invalid .tex document, and tinytex failed to compile it.

> Originally posted on 2019-08-14 03:18:30

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Try to uninstall MiKTeX and install TinyTeX instead: <https://bookdown.org/yihui/rmarkdown/installation.html>

> Originally posted on 2019-09-04 04:17:53

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
That's great to know! Thanks for the kind words, Roger! (this name always reminds me of quantile regression... :)

> Originally posted on 2019-12-11 20:37:13

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
You can safely ignore the messages like "/usr/local/... not writable". They are harmless. If tinytex:::is_tinytex() doesn't return TRUE, it must have been uninstalled successfully.

> Originally posted on 2020-03-03 15:04:38

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I don't understand these error/warning messages, either, but you may try tinytex:::install_prebuilt().

> Originally posted on 2020-03-03 15:05:35

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I have changed the Github URLs. With the latest CRAN version of tinytex, you should be able to install TinyTeX in China. Thanks!

> Originally posted on 2020-06-04 02:14:34

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
See FAQ 3: <https://yihui.org/tinytex/faq/>

> Originally posted on 2020-06-04 02:13:09

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sounds like you were using a very old version of the R package tinytex. After you install the Github version, did you restart your R session?

> Originally posted on 2020-06-09 14:27:57

**Reply by yihui** [reply]
You only need to restart once. Installing from CRAN should suffice (no need to install from Github, although it's harmless). What's your xfun::session_info('tinytex')?

> Originally posted on 2020-06-10 19:44:45

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
It depends on what you mean by "update". It could be either tinytex::tlmgr_update() or tinytex::reinstall_tinytex(). For the latter, the installed packages will be reinstalled, but none of your settings will be kept (I'm not sure what kind of settings you meant).

> Originally posted on 2020-07-25 14:13:29

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
First, you are not using TinyTeX but Ubuntu's TeX Live. Second, I don't think your R packages are up-to-date: <https://yihui.org/tinytex/r/#debugging>

> Originally posted on 2020-07-30 00:32:37

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
What is the exact command that you ran? Chances are you missed the backticks around whoami: <https://github.com/yihui/tinytex/issues/24#issuecomment-372043672>

> Originally posted on 2020-12-30 05:46:10

**Reply by yihui** [reply]
You should not have typed the command manually. Please copy and paste it from the link I mentioned above. In your case, you typed single quotes instead of backticks, which completely changed the meaning of the command.

> Originally posted on 2020-12-30 20:33:14

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
This RStudio issue has been resolved recently. Please see <https://github.com/yihui/tinytex/issues/10#issuecomment-764648243> and try the daily build of RStudio.

> Originally posted on 2021-02-11 23:23:29

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
抱歉 MiKTeX 的问题我无法解答，你可以试试 TinyTeX；安装方法请往上翻到顶。

> ——原帖发布于 2021-03-22 05:24:35

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
To update R packages, you can run update.packages(): <https://yihui.org/tinytex/r/#debugging>

> Originally posted on 2021-03-30 15:45:59

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
No, you don't.

> Originally posted on 2021-09-05 13:07:34

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry, but I can't help before you read and follow the debugging tips in the error message. The information you provided was too little.

> Originally posted on 2021-09-09 20:39:23

[a comment by an automated account is not shown]

**Comment by kjayhan** [comment]
I use Rstudio on my both Mac and Windows (files on the cloud).

While I am able to render my Quarto file (a journal article draft) into a pdf on my Mac, it does not render on Windows. The error message is as follows:

"running xelatex - 1
This is XeTeX, Version 3.141592653-2.6-0.999993 (TeX Live 2021/W32TeX) (preloaded format=xelatex)
restricted \write18 enabled.
entering extended mode

updating tlmgr

updating existing packages
finding package for setspace.sty

compilation failed- no matching packages
LaTeX Error: File `setspace.sty' not found.

Type X to quit or to proceed,
or enter new name. (Default extension: sty)

Enter file name:
! Emergency stop.
<read *>

l.14 \usepackage
{iftex}^^M
"

The following and other related advice did not work:

tinytex::parse_install(here::here("myfile.log")).

$ tlmgr search --global --file "/setspace.sty"
psnfss:
        texmf-dist/tex/latex/psnfss/setspace.sty


How can I work around this?

**Reply by Teddy-van-Jerry** [reply]
Hi, in the error message, it seems to be complaining about `\usepackage{iftex}`. Have you checked whether you have that installed?

**Reply by yihui** [reply]
@kjayhan It seems you are using an old version of TinyTeX (2021). I recommend that you upgrade or reinstall TinyTeX first.

**Reply by kjayhan** [reply]
I updated ALL packages including TinyTex, but it didn't help. I will try installing iftex tomorrow (when I have access to my Windows computer).

**Reply by kjayhan** [reply]
I reinstalled "tinytex", but couldn't install "iftex" since the latter was not available for my R version (4.2.1).

For tinytext, I got this message:   

There is a binary version available but the source version is later:
        binary source needs_compilation
tinytex   0.43   0.44             FALSE

**Reply by yihui** [reply]
I meant TinyTeX, not tinytex: https://yihui.org/tinytex/#for-r-users

To reinstall TinyTeX, see: https://yihui.org/tinytex/r/#debugging

BTW, `iftex` is a LaTeX package, not an R package. You can't use `install.packages()` to install it. And I'm not sure if @Teddy-van-Jerry's guess was correct. In theory, you shouldn't need to deal with any missing LaTeX packages when you are using TinyTeX: https://yihui.org/tinytex/r/#compile-latex-documents

**Reply by kjayhan** [reply]
Thank you for the suggestions. I will try running them on my Windows next time I have access to it.

For the time being I have a similar problem both on my Mac and Windows.

I am using this template for writing a manuscript: https://github.com/ajnafa/project-template

It always tells me 

> compilation failed- package installation error
> Package fontspec Error: The font "FontAwesome" cannot be found.

I did everything on the links you sent (installing TinyTex, updating packages, tinytex::tlmgr_update() etc.), tinytex::parse_install, as well as getting the extension for FontAwesome (in addition to the R package).

Nothing solves the problem (except removing fontawesome lines from Manuscript_Template.tex file (which leaves me without the author names etc.).

I am a novice to working with Latex files as you can tell, so the learning curve is steep to change even a font (especially when it involves -orcid, github etc.- logos).

**Reply by kjayhan** [reply]
Reinstalling TinyTex did solve the "setspace.sty" problem on my Windows, thank you. 

But Macbook still cannot find FontAwesome which may be because the author of[ that template](https://github.com/ajnafa/project-template) actually uses MikTex.

**Comment by Godric-Wang** [comment]
Hi Yihui,

我是tinytex的mac用户，最近我想为公司的一台电脑安装tinytex，但是公司非常注重数据的安全问题，目前电脑并没有也不允许连接网络，我已经通过password protected的USB成功安装了R和Rstudio，和里面需要的必要package， 但是tinytex是一个LaTex的distribution，你提供的下载和安装方式要么是通过R studio或者通过bash或者powershell联网实现，想问一下我能否在自己的mac上下载windows版本的TinyTex，然后通过USB转移进去后离线安装么？可以提供一下我这种情况的详细教程吗？公司电脑是windows版本

提前感谢你的解答！

**Reply by yihui** [reply]
TinyTeX 是可移植的，所以自然可以用 USB 拷贝。参见常见问题 9：https://yihui.org/tinytex/faq/#faq-9

**Comment by lingo555** [comment]
你好，我不确定这个问题是否应该问您，但找不到地方问。

我在Rstudio上使用Rmarkdown，想输出pdf文档于是选用了tinytex。但是目前的状况是：
(1)我在日本工作，文档里有日语，网上查了下，得知通过“tinytex::tlmgr_install("ipaex")”安装IPAex字体即可（实际上我自己的私人电脑也试过可行）
(2)然而我公司因为安全问题无法连外网，自然也无法通过上面方法安装

于是想知道有没有什么方法（就像您提供的本地安装tinytex的方法那样）可以本地安装这个字体吗？

如果您不了解的话我就干干脆脆放弃了，也问题不大。
谢谢！

**Reply by yihui** [reply]
可以。不能连外网没有问题，只要能拷贝文件进公司电脑即可；要是不能拷贝，那么只要允许用 U 盘即可：https://yihui.org/en/2018/08/tinytex-flash-drive/
